hogben/models/base.py

```diff
@@ -56,16 +55,61 @@ class BaseSample(VariableAngle):
     """Abstract class representing a "standard" neutron reflectometry sample
     defined by a series of contiguous layers."""
 
-    @abstractmethod
-    def sld_profile(self):
-        """Plots the SLD profile of the sample."""
-        pass
+    def __init__(self):
+        """Initialise the sample and define class attributes"""
+        self._structures = []
 
     @abstractmethod
-    def reflectivity_profile(self):
-        """Plots the reflectivity profile of the sample."""
+    def get_structures(self):
+        """
+        Get a list of the possible sample structures.
+        """
         pass
 
+    @property
+    def structures(self):
+        """Return the structures that belong to the sample"""
+        return self.get_structures()
+
+    @structures.setter
+    def structures(self, structures):
+        self._structures = structures
+
+    def get_param_by_attribute(self, attr: str) -> list:
+        """
+        Get all parameters defined in the sample model that have a given
+        attribute. Returns a list with all parameters with this attribute,
+        e.g. `attr='vary'` returns all varying parameters.
+
+        Args:
+            attr (str): The attribute to filter for
+
+        Returns:
+            list: A list of all parameters with the given attribute
+        """
+        params = []
+        for model in self.get_models():
+            for p in flatten(model.parameters):
+                if hasattr(p, attr) and getattr(p, attr):
+                    params.append(p)
+                    continue
+                if p._deps:
```

Comment:
This is basically an adaption from the `varying_parameters` method in `refnx` ([link](https://github.com/refnx/refnx/blob/2fd237d10a8b5c065c186837e7e40821e1219be0/refnx/analysis/objective.py#L372)), but extended to allow for "custom" attributes as well instead of just the `vary` parameter:

```

    def varying_parameters(self):
        """
        Returns
        -------
        varying_parameters : refnx.analysis.Parameters
            The varying Parameter objects allowed to vary during the fit.

        """
        # create and return a Parameters object because it has the
        # __array__ method, which allows one to quickly get numerical values.
        lst = []
        for p in flatten(self.parameters):
            if p.vary:
                lst.append(p)
                continue
            if len(p._deps):
                lst.extend([_p for _p in p.dependencies() if _p.vary])
        # should already be totally flattened by this point
        return Parameters(f_unique(lst))
        ```
        
See also `p.dependencies`:

```

    def dependencies(self):
        dep_list = []
        for _dep in self._deps:
            if isinstance(_dep, Parameter):
                if _dep.constraint is not None:
                    dep_list.append(_dep.dependencies())
                else:
                    dep_list.append(_dep)
            if isinstance(_dep, (_UnaryOp, _BinaryOp)):
                dep_list.append(_dep.dependencies())
        return list(f_unique(flatten(dep_list)))
```

Where the second statement checks the nested dependencies, and adds these as well if the attributes match. I'll have to do a double check, but this could be for instance the iSLD and the real SLD. Will clarify a bit in the code itself.

Reply:
Did some investigation, and this indeed gets the parameters that are coupled to the model due to dependencies. So to clarify, imagine our sample is like: `substrate | layer1| layer2| air|`, and we've got the following parameters:
```
    gamma = Parameter(0.5, 'Gamma', (0, 1))
    period = Parameter(120, 'Period', (60, 190))
    layer1_thick = period*gamma
    layer2_thick = period*(1-gamma)
    gamma.vary = True
```

The `gamma` parameter will actually be some `BinaryOps` parameter that will not have the `vary` attribute. But this parameter will instead be part of the `parameter.dependencies` attribute. So what this second `if` basically does is adding the parameters that are not directly tied to an SLD-slab parameter, but instead tied to the model using such dependencies as the code-block in this comment above. I added a short comment here to clarify this somewhat.